File: Args.cs
namespace IPK_2nd_project;

public class Args
{
    public ushort PortSource;
    public ushort PortDestination;
    public ushort Port;
    public bool Tcp;
    public bool Udp;
    public string? Interface;
    public int NumberOfPackets;
    public bool Icmp4;
    public bool Icmp6;
    public bool Arp;
    public bool Ndp;
    public bool Igmp;
    public bool Mld;


    public void Parse(string[] args)
    {
        Tcp = false;
        Udp = false;
        Icmp4 = false;
        Icmp6 = false;
        Arp = false;
        Ndp = false;
        Igmp = false;
        Mld = false;
        NumberOfPackets = 1;
        Interface = null;
        
        if (args.Length == 0)
        {
            Stderr.Write("No arguments");
            return;
        }

        // ./ipk-sniffer [-i interface | --interface interface] {-p|--port-source|--port-destination port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] {-n num}
        for (int index = 0; index < args.Length; index++)
        {
            string current = args[index];
            if (current == "--port-destinations" || current == "--port-source" || current == "-p")
            {
                if (index + 1 < args.Length)
                {
                    if (current == "--port-destinations")
                    {
                        PortDestination = ushort.Parse(args[++index]);
                    }
                    else if (current == "--port-source")
                    {
                        PortSource = ushort.Parse(args[++index]);
                    }
                    else
                    {
                        Port = ushort.Parse(args[++index]);
                    }
                }

                if (index + 1 < args.Length)
                {
                    if (args[++index] == "--tcp" || args[index] == "-t")
                    {
                        Tcp = true;
                    }
                    else if (args[index] == "--udp" || args[index] == "-u")
                    {
                        Udp = true;
                    }
                }
            }
            else if (current == "-i" || current == "--interface")
            {
                if (index + 1 < args.Length)
                {
                    Interface = args[++index];
                }
            }
            else if (current == "--icmp4")
            {
                Icmp4 = true;
            }
            else if (current == "--icmp6")
            {
                Icmp6 = true;
            }
            else if (current == "--arp")
            {
                Arp = true;
            }
            else if (current == "--ndp")
            {
                Ndp = true;
            }
            else if (current == "--igmp")
            {
                Igmp = true;
            }
            else if (current == "--mld")
            {
                Mld = true;
            }
            else if (current == "-n")
            {
                if (index + 1 < args.Length)
                {
                    NumberOfPackets = int.Parse(args[++index]);
                }
            }
            else
            {
                Stderr.Write("Wrong argument");
            }
        }
    }
}

File: Args.cpp
#include "Args.h"
/**
 * @brief Constructor of object of class Args
*/
Args::Args()
{
    tcp = false;
    udp = false;
    icmp4 = false;
    icmp6 = false;
    arp = false;
    ndp = false;
    igmp = false;
    mld = false;
    numberOfPackets = 1;
    interface = "";
    port = -1;
    portSource = -1;
    portDestination = -1;
}

/**
 * @brief Parse the arguments from the command line\
 * 
 * @param argc number of arguments
 * @param argv array of arguments
*/
void Args::parse(int argc, char* argv[])
    {
        for (int i = 1; i < argc; i++) // start from 1 to skip the program name
        {
            string arg = argv[i];
            if (arg == "--port-destination")
            {
                if (port != -1) // if port is set, then it is not possible to set port destination
                {
                    handleError("Error: port destination is already set");
                }
                if (i + 1 < argc)
                {
                    portDestination = atoi(argv[i + 1]);
                    i++;
                }
                else
                {
                    handleError("Error: port destination is missing");
                }
            }
            else if (arg == "--port-source")
            {
                if (port != -1) // if port is set, then it is not possible to set port source
                {
                    handleError("Error: port source is already set");
                }
                if (i + 1 < argc)
                {
                    portSource = atoi(argv[i + 1]);
                    i++;
                }
                else
                {
                    handleError("Error: port source is missing");
                }
            }
            else if (arg == "--port" || arg == "-p")
            {
                if (portSource != -1 || portDestination != -1) // if port source or port destination is set, then it is not possible to set port
                {
                    handleError("Error: ports are already set");
                }
                if (i + 1 < argc)
                {
                    port = atoi(argv[i + 1]);
                    i++;
                }
                else
                {
                    handleError("Error: port is missing");
                }
            }
            else if (arg == "--interface" || arg == "-i")
            {
                if (i + 1 < argc)
                {
                    interface = argv[i + 1];
                    i++;
                }
                else
                {
                    handleError("Error: interface is missing");
                }
            }
            else if (arg == "--number-of-packets" || arg == "-n")
            {
                if (i + 1 < argc)
                {
                    numberOfPackets = atoi(argv[i + 1]);
                    i++;
                }
                else
                {
                    handleError("Error: number of packets is missing");
                }
            }
            else if (arg == "--tcp" || arg == "-t")
            {
                tcp = true;
            }
            else if (arg == "--udp" || arg == "-u")
            {
                udp = true;
            }
            else if (arg == "--icmp4")
            {
                icmp4 = true;
            }
            else if (arg == "--icmp6")
            {
                icmp6 = true;
            }
            else if (arg == "--arp")
            {
                arp = true;
            }
            else if (arg == "--ndp")
            {
                ndp = true;
            }
            else if (arg == "--igmp")
            {
                igmp = true;
            }
            else if (arg == "--mld")
            {
                mld = true;
            }
            else if (arg == "--help" || arg == "-h")
            {
                printHelp();
            }
            else
            {
                handleError("Error: invalid argument");
            }
        }
    }